rust/worker/src/execution/operators/metadata_filtering.rs

```diff
@@ -0,0 +1,1348 @@
+                            .collect();
+                        Some(ids_as_u32)
+                    }
+                    None => None,
+                }
+            }
+            Err(e) => {
+                return Err(MetadataFilteringError::MetadataFilteringMetadataSegmentReaderError(e));
+            }
+        };
+        // It cannot happen that one is none and other is some.
+        if (filter_from_mt_segment.is_some() && merged_result.is_none())
+            || (filter_from_mt_segment.is_none() && merged_result.is_some())
+        {
+            panic!("Invariant violation. Both should either be none or some");
+        }
+        let mut where_condition_filtered_offset_ids = None;
+        if filter_from_mt_segment.is_some() && merged_result.is_some() {
+            where_condition_filtered_offset_ids = Some(merge_sorted_vecs_disjunction(
+                filter_from_mt_segment.expect("Already checked that should be some"),
+                merged_result.expect("Already checked that should be some"),
+            ));
+        }
+
+        // Hydrate offset ids for user supplied ids.
+        // First from the log.
+        let mut user_supplied_offset_ids: Vec<u32> = vec![];
+        let mut remaining_id_set: HashSet<String> = HashSet::new();
+        let mut query_ids_present = false;
+        match &input.query_ids {
+            Some(query_ids) => {
+                let query_ids_set: HashSet<String> = HashSet::from_iter(query_ids.iter().cloned());
+                query_ids_present = true;
+                remaining_id_set = query_ids.iter().cloned().collect();
+                for (log_records, _) in mat_records.iter() {
+                    match log_records.user_id {
+                        Some(id) => {
+                            if query_ids_set.contains(id) {
+                                remaining_id_set.remove(id);
+                                if log_records.final_operation != Operation::Delete {
+                                    user_supplied_offset_ids.push(log_records.offset_id);
+                                }
+                            }
+                        }
+                        None => match &log_records.data_record {
+                            Some(data_record) => {
+                                if query_ids_set.contains(data_record.id) {
+                                    remaining_id_set.remove(data_record.id);
+                                    if log_records.final_operation != Operation::Delete {
+                                        user_supplied_offset_ids.push(log_records.offset_id);
+                                    }
+                                }
+                            }
+                            None => panic!("Invariant violation. Expected at least one user id to be set in materialized record"),
+                        },
+                    }
+                }
+                let record_segment_reader_2: Option<RecordSegmentReader>;
```

Comment:
why do we create a 2?

Reply:
Because the materializer owns the reader 1. And clone() is not allowed on readers

Reply:
ah - can we just add a clone?